File: js/services/invoice-service.js
import { database } from '../firebase-config.js';
import { ref, push, set, get, remove, query, orderByChild, limitToLast } from 'firebase/database';

export class InvoiceService {
    constructor() {
        this.path = 'invoices';
    }

    async create(invoice) {
        const invoiceRef = ref(database, this.path);
        const newRef = push(invoiceRef);
        await set(newRef, invoice);
        return newRef.key;
    }

    async update(id, invoice) {
        const invoiceRef = ref(database, `${this.path}/${id}`);
        await set(invoiceRef, invoice);
    }

    async delete(id) {
        const invoiceRef = ref(database, `${this.path}/${id}`);
        await remove(invoiceRef);
    }

    async get(id) {
        const invoiceRef = ref(database, `${this.path}/${id}`);
        const snapshot = await get(invoiceRef);
        if (snapshot.exists()) {
            return { id: snapshot.key, ...snapshot.val() };
        }
        return null;
    }

    async getAll() {
        const invoicesRef = ref(database, this.path);
        const snapshot = await get(invoicesRef);
        const invoices = [];
        if (snapshot.exists()) {
            snapshot.forEach(child => {
                invoices.push({ id: child.key, ...child.val() });
            });
        }
        return invoices;
    }

    async generateInvoiceNumber() {
        const invoicesRef = ref(database, this.path);
        const q = query(invoicesRef, orderByChild('invoiceNumber'), limitToLast(1));
        const snapshot = await get(q);
        
        let nextNumber = 1;
        if (snapshot.exists()) {
            snapshot.forEach(child => {
                const invoice = child.val();
                if (invoice.invoiceNumber) {
                    const match = invoice.invoiceNumber.match(/\d+/);
                    if (match) {
                        nextNumber = parseInt(match[0]) + 1;
                    }
                }
            });
        }
        
        return `INV-${String(nextNumber).padStart(4, '0')}`;
    }
}

export const invoiceService = new InvoiceService();

File: lib/services/invoice_service.dart
import '../models/invoice.dart';
import 'firebase_service.dart';

class InvoiceService {
  final FirebaseService _firebase;
  final String _path = 'invoices';

  InvoiceService(this._firebase);

  Future<String> create(Invoice invoice) async {
    return await _firebase.save(_path, invoice.toJson());
  }

  Future<void> update(Invoice invoice) async {
    if (invoice.id == null) throw Exception('Invoice ID is required');
    await _firebase.update(_path, invoice.id!, invoice.toJson());
  }

  Future<void> delete(String id) async {
    await _firebase.delete(_path, id);
  }

  Future<Invoice?> get(String id) async {
    final data = await _firebase.get(_path, id);
    return data != null ? Invoice.fromJson(data, id) : null;
  }

  Future<List<Invoice>> getAll() async {
    final data = await _firebase.getAll(_path);
    return data.map((item) => Invoice.fromJson(item, item['id'])).toList();
  }

  Future<String> generateInvoiceNumber() async {
    final invoices = await getAll();
    int nextNumber = 1;
    
    if (invoices.isNotEmpty) {
      final numbers = invoices
          .map((inv) {
            final match = RegExp(r'\d+').firstMatch(inv.invoiceNumber);
            return match != null ? int.tryParse(match.group(0)!) : 0;
          })
          .where((n) => n != null && n > 0)
          .cast<int>() // Explicit cast to ensure type safety
          .toList();
      
      if (numbers.isNotEmpty) {
        nextNumber = numbers.reduce((a, b) => a > b ? a : b) + 1;
      }
    }
    
    return 'INV-${nextNumber.toString().padLeft(4, '0')}';
  }
}


